File: Tortuga/Source/Tortuga/Graphics/VulkanAPI/Semaphore.cpp
#include "Semaphore.h"

namespace Tortuga
{
namespace Graphics
{
namespace VulkanAPI
{
Semaphore::Semaphore(Device *device)
{
    this->_device = device;

    auto semaphoreInfo = VkSemaphoreCreateInfo();
    semaphoreInfo.sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO;

    if (vkCreateSemaphore(_device->GetVirtualDevice(), &semaphoreInfo, nullptr, &_semaphore) != VK_SUCCESS)
    {
        Console::Fatal("Failed to create semaphore for command buffer!");
    }
}

Semaphore::~Semaphore()
{
    vkDestroySemaphore(_device->GetVirtualDevice(), _semaphore, nullptr);
}
}; // namespace VulkanAPI
}; // namespace Graphics
}; // namespace Tortuga

File: Tortuga.Graphics/API/Semaphore.cs
#pragma warning disable CS1591
using System;
using Vulkan;

namespace Tortuga.Graphics.API
{
    public class Semaphore
    {
        public Device Device => _device;
        public VkSemaphore Handle => _handle;

        private Device _device;
        private VkSemaphore _handle;

        public unsafe Semaphore(Device device)
        {
            _device = device;
            var createInfo = new VkSemaphoreCreateInfo
            {
                sType = VkStructureType.SemaphoreCreateInfo
            };

            VkSemaphore semaphore;
            if (VulkanNative.vkCreateSemaphore(
                _device.Handle,
                &createInfo,
                null,
                &semaphore
            ) != VkResult.Success)
                throw new Exception("failed to create semaphore");
            _handle = semaphore;
        }
        unsafe ~Semaphore()
        {
            if (_handle != VkSemaphore.Null)
            {
                VulkanNative.vkDestroySemaphore(
                    _device.Handle,
                    _handle,
                    null
                );
                _handle = VkSemaphore.Null;
            }
        }
    }
}